CenterlineMetrics/CenterlineMetrics.py

```diff
@@ -382,11 +443,91 @@ def getExtremeDiameterPoint(self, boolMaximum):
             break
     return target
 
-  def getUnitNodeDisplayString(self, value):
+  def getUnitNodeDisplayString(self, value, category):
     selectionNode = slicer.mrmlScene.GetNodeByID("vtkMRMLSelectionNodeSingleton")
-    unitNode = slicer.mrmlScene.GetNodeByID(selectionNode.GetUnitNodeID("length"))
+    unitNode = slicer.mrmlScene.GetNodeByID(selectionNode.GetUnitNodeID(category))
     return unitNode.GetDisplayStringFromValue(value)
     
+  
+  def createCrossSection(self, center, normal):
+    if self.segmentationNode is None or self.currentSegmentID == "":
+        # Don't print any message, is perhaps intended.
+        return
+    
+    util = Utility()
+    util.islandModelNode = self.islandModelNode
+    
+    # Place a plane perpendicular to the centerline
+    plane = vtk.vtkPlane()
+    plane.SetOrigin(center)
+    plane.SetNormal(normal)
+    
+    # Work on the segment's closed surface
+    closedSurfacePolyData = vtk.vtkPolyData()
+    self.segmentationNode.GetClosedSurfaceRepresentation(self.currentSegmentID, closedSurfacePolyData)
+    
+    # Though not needed here, it was surprising that vtkSegment::GetId() is missing.
+    currentSegment =  self.segmentationNode.GetSegmentation().GetSegment(self.currentSegmentID)
+    
+    # Cut through the closed surface and get the points of the contour.
+    planeCut = vtk.vtkCutter()
+    planeCut.SetInputData(closedSurfacePolyData)
+    planeCut.SetCutFunction(plane)
+    planeCut.Update()
+    planePoints = vtk.vtkPoints()
+    planePoints = planeCut.GetOutput().GetPoints()
+    # self.segmentationNode.GetDisplayNode().GetSegmentVisibility3D(self.currentSegmentID) doesn't work as expected. Even if the segment is hidden in 3D view, it returns True.
+    if planePoints is None:
+        msg = "Could not cut segment. Is it visible in 3D view?"
+        slicer.util.showStatusMessage(msg, 3000)
+        raise ValueError(msg)
+    if planePoints.GetNumberOfPoints() < 3:
+        logging.info("Not enough points to create surface")
+        return
+    
+    # Sort the points by proximity to retain those around the centre of the centerline
+    vCenter = [center[0], center[1], center[2]]
+    planePointsCopy = vtk.vtkPoints()
+    planePointsCopy.DeepCopy(planePoints)
+    # Empties planePointsCopy
+    sortedPoints = util.sortPointsByProximity(planePointsCopy, vCenter)
```

Comment:
Could you use vtkConnectivityFilter's ExtractionModeToClosestPointRegion method instead?

https://vtk.org/doc/nightly/html/classvtkConnectivityFilter.html#a0afa68eb6a4aded83fbc8c0951ac5a69

Reply:
> vtkConnectivityFilter's ExtractionModeToClosestPointRegion method

It is now used to select the island around the centerline. But the points still need to be rearranged by proximity, else the code becomes a CPU burner !

Reply:
The slice intersection should not contain more than a thousand points (more likely a few hundred). Any C++ code would do any processing on such small number of points in a fraction of a millisecond. Do you mean that there is still some processing that you would need to do in Python?

Reply:
> should not contain more than a thousand points

The maximum number of points per section is 168 in my test data.

If I just pass the output of connectivityFilter.GetPolyDataOutput().GetPoints() to FitSurfaceProjectWarp(), a deadly loop is triggered, eating all RAM and CPU. If the points are sorted by proximity, then the model is fastly created.

Now you advise to use vtkContourTriangulator instead of FitSurfaceProjectWarp(). That's perhaps the issue. I'll try that and get back.

Reply:
FitSurfaceProjectWarp expects the points in order. If you are lucky then sorting by distance is similar enough, but in general you cannot rely on this. Probably if you create models from the cross-sections then you'll see errors if you order polygon points based on distance. Anyway, vtkContourTriangulator takes care of this, too, as it uses connectivity between points to order them correctly.

Reply:
I understand the advantages of using vtkContourTriangulator.
I am unable to get a surface model with : 

contourTriangulator = vtk.vtkContourTriangulator()
islandPolyData.SetPoints(islandPoints)
islandPolyData.Modified()
contourTriangulator.SetInputData(islandPolyData)
contourTriangulator.Update()
slicer.modules.models.logic().AddModel(contourTriangulator.GetOutput())

Calling contourTriangulator.Triangulate is worse, leading to a crash.

Reply:
Contour triangulator requires the entire cross-section of the polygon, not just the points.

Reply:
> entire output of the cutter (including cells - connectivity between points), not just the points
This did the trick, now it works, thanks.